core/serialisation/src/main/java/uk/gov/gchq/gaffer/serialisation/ToBytesViaStringDeserialiser.java

```diff
@@ -56,6 +57,9 @@ public final T deserialise(final byte[] bytes) throws SerialisationException {
     @Override
     public final T deserialise(final byte[] allBytes, final int offset, final int length) throws SerialisationException {
         try {
+            if (Arrays.equals(Arrays.copyOfRange(allBytes, offset, offset + length), NULL_BYTES)) {
+                return null;
```

Comment:
This isn't really the kind of multiple return that makes the code harder to read, I actually think it is easier to read and [less verbose](https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html). Debugging would also be the same as the breakpoint would not be hit if inside an else block.

Reply:
I disagree.
The example you provided is talking about using multiple `else` then `if` statements. 
You don't have that here. The first example is horrendous and verbose anything will look better than that ugly first example. You'd use `if else` on a single line and you still require the encapsulating braces for CheckStyle in our project.

```java 
function getPayAmount() {
 rtn ;
  if (isDead) { rtn = deadAmount(); }
  else if (isSeparated) { rtn =  separatedAmount(); }
  else if (isRetired) { rtn =  retiredAmount(); }
  else { rtn =  normalPayAmount(); }
  return rtn;
 //single return 
}
```


 This code being examined in our code is single `If` statement with no nesting. Multiple returns provide little to no benefit here.

4 parts & 2 returns which at times would need a debug break point on each return.
```java
// 1 if statement
    if (Arrays.equals(Arrays.copyOfRange(allBytes, offset, offset + length), NULL_BYTES)) {
// 2 return statement
         return null;
     }
// 3 assignment
             String valueString = new String(allBytes, offset, length, charset);
// 4 return statements
             return deserialiseString(valueString);
}             
```


vs 1 return for 1 `if` statement 
```java
T rtn = null;

// 1 if statement
    if (!Arrays.equals(Arrays.copyOfRange(allBytes, offset, offset + length), NULL_BYTES)) {
// 2 assignment
      rtn =  deserialiseString(new String(allBytes, offset, length, charset));
    }
// 3 return
return  rtn
```

This follows a Edsger Dijkstra’s approach of having 1 exit from a method/function.

if you really want to just smash in a ternary statement. can't get less verbose than that.
```java
return  (!Arrays.equals(Arrays.copyOfRange(allBytes, offset, offset + length), NULL_BYTES)) 
    ? deserialiseString(new String(allBytes, offset, length, charset))
    : null;
```

Reply:
Single exit code style is [outdated](https://softwareengineering.stackexchange.com/questions/118703/where-did-the-notion-of-one-return-only-come-from) and I still think the example you provide for the `getPayAmount` function still looks worse and won't work nicely in a debugger.
Having said that, I do think that flipping the if statement like you did makes this function a lot cleaner so I will amend it. Cheers for the feedback!

Reply:
> example you provide for the getPayAmount function still looks worse and won't work nicely in a debugger.

It has more characters to allow the single return type yes.

But with the example given when you scanning through code to find where an unusual value is being created
you would require 4 break points and you'd need to step/in/over to view the value. The debug stops on line ` if (isSeparated) { return separatedAmount(); }` but you don't know the value of this line, you gotta manipulate the debug step in/out to get the value of `separatedAmount()`

```Java
function getPayAmount() { 
  if (isDead) { return deadAmount(); } //break 1
  if (isSeparated) { return separatedAmount(); } //break 2
  if (isRetired) { return retiredAmount(); } //break 3
  return { normalPayAmount(); } //break 4
}
```

with the single return the breakpoint, it will have already resolved the value of this method `rtn` in a single hit at the breakpoint, without needing to tab in/tab out of methods to resolve on the breakpoint line. The `rtn` value is visible and ready
```java
function getPayAmount() {
 rtn ;
  if (isDead) { rtn = deadAmount(); }
  else if (isSeparated) { rtn =  separatedAmount(); }
  else if (isRetired) { rtn =  retiredAmount(); }
  else { rtn =  normalPayAmount(); }
  return rtn; //break 1 !!!!!!
}
```

^ this is why I like to write like this I think it looks cleaners and debugging has less steps and tabbing around.

yeah Edsger Dijkstra’s is super old school. 
I like it as an approach to writing as it is usually cleaner to have 1 return partially for what I perceive/describe above^

> "Structured Programming
> Some programmers follow Edsger Dijkstra’s rules of structured programming.14 Dijkstra said that every function, and every block within a function, should have one entry and one exit. Following these rules means that there should only be one return statement in a func- tion, no break or continue statements in a loop, and never, ever, any goto statements.
> While we are sympathetic to the goals and disciplines of structured programming, those rules serve little benefit 
> **when functions are very small. It is only in larger functions that such rules provide significant benefit.
> So if you keep your functions small, then the occasional multiple return, break, or continue statement does no harm and can sometimes even be more expressive than the sin- gle-entry, single-exit rule.** 
> " [Clean Code: A Handbook of Agile Software Craftsmanship (Robert C. Martin) (Robert C. Martin Series)]

I don't see a an advantage to having 4 returns here, when 1 will do.

This has been/become an exercise/discussion programming approaches. 

You can keep this code as you like. But I hope you understand the debug breakpoint reasoning I use/have.

When your in Debug Hell you want every little morsel on your side :D

Reply:
I see what you mean about debugging and in large functions like the quote says, I think this approach can be helpful. But in 4 line functions I think debugging is easy enough. You are right though it is preference at this point. I did modify it though to swap the if statement as it looks nicer that way.